README.md
# node-red-contrib-spark

[Node-RED](http://nodered.org) Nodes to integrate with the [Cisco Webex Teams API](https://developer.webex.com).

Version 2.0.0 [(changelog)](https://github.com/cumberlandgroup/node-red-contrib-spark/blob/master/CHANGELOG.md)

![](https://github.com/cumberlandgroup/node-red-contrib-spark/raw/master/images/flow01.jpg)

### Nodes

* **api** - This Node receives queries on the input that are then sent to the Webex Teams API. The results are provided at the output.
* **webhook** - This Node creates, removes, and manages the Webhook features of the Webex Teams API. Once deployed, and subsequently triggered, the contents of the notification will be sent to the output of this Node.
* **parser** - This is a utility Node that accepts input from the output of either the API or Webhook Node.
* **auth** - This is a Config Node that holds the credential Bearer Token for the Webex Teams API. Once initially defined, either under the Api or Webhook Node, it will be available to select on all other Api and Webhook Nodes that are created. You can define multiple Auth profiles so as to work with different Webex Teams Accounts or Bots within the same flow.

### Install

The simplest installation method is done via Node-RED itself. In the "Manage Palette" menu dropdown, search for this module by name. Alternate methods are outlined below.

**Via the node-red-admin CLI tool:**

```bash
# requires node-red to be running

# install node-red-admin if needed
npm install -g node-red-admin

# authenticate if your Node-RED administration has been secured
node-red-admin login

# install the module
node-red-admin install node-red-contrib-spark
```

**Via source:**
```bash
# clone repo
git clone https://github.com/cumberlandgroup/node-red-contrib-spark
cd node-red-contrib-spark

# get submodule
git submodule init
git submodule update

# from your node-red user directory
cd $HOME/.node-red
npm install /path/to/node-red-contrib-spark
```

**Via NPM repository:**
```bash
# from your node-red user directory
cd $HOME/.node-red
npm install node-red-contrib-spark
```

## API Node

The Webex Teams API Node sends REST queries via messages received by the input connector in the `msg.payload` object. Results of the API call are provided at the output in the `msg.payload` object. If multiple records are returned from the Webex Teams API Call, these are passed to the output as individual sequential messages. The `msg.parts` property is set appropriately for use with the `join` node if a single array payload is preferred.

![](https://github.com/cumberlandgroup/node-red-contrib-spark/raw/master/images/api-node.jpg)

#### Module Input

The Webex Teams API request is passed as a JSON object in `msg.payload`. The `msg.payload` object can be an empty object `{}` or contain optional path, query-string, or body variables provided using `"key": "value"` object parameters. Depending on the particular API method, the `"key": "value"` properties are defined in either the `msg.payload` object or the `msg.payload.body` object.

#### Module Output

By convention, the output from the Webex Teams API call will have a `msg.payload` property. This contains the results of the API call in JSON format. The format of this JSON object will be the same as documented at [developer.webex.com](https://developer.webex.com) for the responses from the API call.

Additionally the following are defined as part of the msg object:

* `msg.status` : http return code
* `msg.headers` - response headers object

#### Multiple Results
If multiple records are returned from the Webex Teams API Call, these are passed to the output as individual sequential messages. The `msg.parts` property is set appropriately for use with the `join` node if a single array payload is preferred.

**Example: Get Person by Email**

The following object would be sent in the `msg.payload` input to a Webex Teams API Node setup for `People.getPeople`:

```json
{
  "email": "person@example.com"
}
```

**Example: Get Rooms by Type**

The following object would be sent in the `msg.payload` input to a Webex Teams API Node setup for `Rooms.getRooms`:

```json
{
  "body": {
    "type": "group"
  }
}
```

**Example: Update a Room Title**

The following object would be sent in the `msg.payload` input to a Webex Teams API Node setup for `Rooms.updateRoom`:

```json
{
  "roomId": "someRoomIdString",
  "body": {
    "title": "My Renamed Room"
  }
}
```

**Example: Create a New Message**

The following object would be sent in the `msg.payload` input to a Webex Teams API Node setup for `Messages.createMessage`:

```json
{
  "body": {
    "roomId": "someRoomIdString",
    "text": "Hello World!"
  }
}
```

**Example: Add Person by Email to a Room**

The following object would be sent in the `msg.payload` input to a Webex Teams API Node setup for `Memberships.createMembership`:

```json
{
  "body": {
    "roomId": "someRoomIdString",
    "personEmail": "person@emaple.com"
  }
}
```

#### Configuration Options

* **Profile** - The Webex Teams credential profile to use with this Node.
* **Resource** - The Webex Teams resources for the API action.
* **Method** - The specific method available to the selected Webex Teams resource.

## Webhook Node

The Webex Teams Webhook Node is triggered when a resource event is matched. When the Node is deployed, it automatically creates the associated Cisco Webex Teams Webhook. When the Node is removed, the Webhook reference is automatically removed in the Webex Teams API.

![](https://github.com/cumberlandgroup/node-red-contrib-spark/raw/master/images/webhook-node.jpg)

**Example Output : `msg.payload`**

```json
{
    "id": "Y2lzY29zcGFyazovL3VzL1dFQkhPT0svNzJkYzlhNTctYmY4MC00OTdjLWFhM2MtNjMyYzUyOThkMTFk",
    "name": "Test",
    "targetUrl": "http://myhost.com:3000/node_red_contrib_spark_388d9fffca49b8",
    "resource": "messages",
    "event": "created",
    "ownedBy": "creator",
    "created": "2016-10-23T19:50:23.484Z",
    "data": {
        "id": "Y2lzY29zcGFyazovL3VzL01FU1NBR0UvZjgyOGM3YTAtOTk1OS0xMWU2LTk5ODYtYzc4MTAwYzIyYTJm",
        "roomId": "Y2lzY29zcGFyazovL3VzL1JPT00vNjNhYzQ3MzAtOTUzYy0xMWU2LWEwZmQtNDcxNWExOWY2ZDJi",
        "roomType": "group",
        "personId": "Y2lzY29zcGFyazovL3VzL1BFT1BMRS8zYzNlZmYxOS04Njg1LTQ2OTEtODViOS1lZjRmMTViZDk2ZDQ",
        "personEmail": "nmarus@gmail.com",
        "created": "2016-10-23T19:50:37.594Z"
    }
}
```

#### Configuration Options

* **Profile** - The Webex Teams credential profile to use with this Node.
* **Resource** - The Webex Teams resource to bind a Webhook to.
* **Event** - The specific event of the resource selected.
* **Host** - The base URL that is used to build the Webhook in the Webex Teams API. This should be reachable from the internet and follow the format of `http(s)://domain.tld:<port>`. The Webhook Node will dynamically publish web routes under this URL as `/node_red_contrib_spark_<node-uuid>`. *Note that the Webhook is automatically created in the Webex Teams API after deploying and automatically removed if the Node is deleted and the flow re-deployed. If you have defined the httpNodeRoot setting in Node-Red, the webhook target URL will be created under that path. This should be mostly transparent as this node will create and remove webhooks from the Webex Teams API as the nodes are added and removed from the workspace. If you shutdown Node-Red with a webhook deployed on the workspace, this webhook will persist in the Webex Teams API and will need to be manually removed.*

## Parser Node

The Webex Teams Parse Node allows parsing of specific properties from the JSON `msg.payload` received from either the "Webex Teams Webhook Node" or the "Webex Teams API Node".

![](https://github.com/cumberlandgroup/node-red-contrib-spark/raw/master/images/parser-node.jpg)

#### Configuration Options

Define each *"property"* to parse from the inbound `msg.payload`. Optionally, define a *"name"* to remap the property name used in the oubound JSON object. This allows remapping of properties such as `msg.payload.id` to `msg.payload.roomId`. If *"name"* is left undefined, it will use the same value as the original property. Add additional rows to define multiple properties and names to construct the output JSON object.

![](https://github.com/cumberlandgroup/node-red-contrib-spark/raw/master/images/parser-node-config.jpg)

## License

MIT License Copyright (c) 2019 Cumberland Group

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 2.1.0: Branding Update

  - Updated mentions of "Spark" to now refer to "Webex Teams"
  - Updated node icon to use the Webex Teams icon
  - Updated README images to reflect branding updates

#### 2.0.5: Maintenance Release

  - Fixed issues with existing API Nodes not loading node form parameters on page refresh.
  - Attempt to fix why nodes are not showing on palette when viewing on nodered.org.
  - Changed references from httpAdmin to httpNode.
  - Fixed issue with webhook node nott showing form elements once a profile is created when webhook node is first node to be initialized.

#### 2.0.4: Maintenance Release

  - Fixed issue #1. Custom httpAdminRoot or httpNodeRoot in settings will now work.
  - Fixed potential race condition between page loading and swagger file being parsed.
  - Tweaked webhook cleanup functions.

#### 2.0.2: Maintenance Release

  - updated api node handling of empty responses

#### 2.0.1: Maintenance Release

  - updated error handling of api node
  - updated swagger definitions

#### 2.0.0: Major Release

  - updated all node inline documentation and label formats
  - api node now sends array responses as individual sequential messages
  - parse node rebuilt to allow more than 1 property to be parsed from payload
  - updated README.md to include updated documentation around parse and api node

#### 1.1.1: Maintenance Release

  - updated README.md typos

#### 1.1.0: Feature Release

  - added msg.topic to api node output that contains api resource and method
  - simplified parser node
  - updated swagger definition files to include new APIs for people, organizations, licenses, and roles

#### 1.0.4: Maintenance Release

  - added msg.topic to parser node output that contains the parser

#### 1.0.3: Maintenance Release

  - updated documentation and corrected typos
  - added parser node

#### 1.0.2: Maintenance Release

  - reduced file sizes
  - changed timing of form population loading after node-red restart and browser refresh
  - added additional help text to auth config node
  - added this CHANGLOG.md file...

#### 1.0.1: Maintenance Release

  - updated node labels to "api" and "webhook"
  - updated node category to "cisco_spark"

#### 1.0.0: Initial Release


spark/spark-webhook.js
module.exports = function(RED) {
  'use strict';
  var bodyParser = require('body-parser');
  var jsonParser = bodyParser.json();
  var request = require('request');
  var path = require('path');
  var fs = require('fs');
  var _ = require('lodash');

  function SparkWebhookNode(n) {
    RED.nodes.createNode(this, n);

    var node = this;

    var httpNodeRoot;

    if(RED.settings.httpNodeRoot) {
      httpNodeRoot = RED.settings.httpNodeRoot;
    } else {
      httpNodeRoot = '/';
    }

    var webhookReq = {
      url: 'https://api.ciscospark.com/v1/webhooks/',
      json: true
    };

    node.profileConfig = RED.nodes.getNode(n.profileConfig);

    node.name = n.name || 'webhhook';
    node.resource = n.resource;
    node.event = n.event;
    node.host = n.host;

    node.basename = 'node_red_contrib_spark_' + node.id.replace(/\./g, '');

    node.pathname = httpNodeRoot + node.basename;
    node.webhookTarget = node.host + node.pathname;

    node.webhookId = n.webhookId || null;
    node.webhookObj = n.webhookObj || null;

    node.reqCount = 0;
    node.reqReceiving = false;

    // set default status
    node.status({
      fill: 'blue',
      shape: 'ring',
      text: 'Webex Teams Webhook: ready'
    });

    // remove webhooks
    node.removeWebhooks = function(callback) {

      // if Webex Teams api token and webhook id are defined...
      if(node.profileConfig.credentials && node.profileConfig.credentials.token) {

        // clone module request template
        var _webhookReq = _.clone(webhookReq);

        // set credentials for request object
        _webhookReq.auth = { bearer: node.profileConfig.credentials.token };

        // send request to delete webhook
        request.get(_webhookReq, function(err, res) {
          // if request error...
          if(err) {
            node.log('error: ' + err.message);
            if(callback) callback(err);
          }

          // else, if webhooks returned...
          else if(res && res.body && res.body.items && res.body.items instanceof Array && node.webhookTarget) {
            _.forEach(res.body.items, function(webhook) {
              if(webhook.targetUrl == node.webhookTarget) {
                node.removeWebhook(webhook.id);
                node.log('Removed existing webhook for: ' + webhook.targetUrl);
              }
            });
            if(callback) callback(null);
          }

          // else...
          else {
            if(callback) callback(null);
          }

        });
      } else {
        if(callback) callback(null);
      }
    };

    // remove a webhook
    node.removeWebhook = function(id) {
      var webhookId = node.webhookId;

      if(typeof id !== 'undefined') {
        webhookId = id;
      }

      // if Webex Teams api token and webhook id are defined...
      if(node.profileConfig.credentials && node.profileConfig.credentials.token && webhookId) {

        // clone module request template
        var _webhookReq = _.clone(webhookReq);

        // set credentials for request object
        _webhookReq.auth = { bearer: node.profileConfig.credentials.token };

        // set delete url for request object
        _webhookReq.url = _webhookReq.url + webhookId;

        // send request to delete webhook
        request.delete(_webhookReq, function(err, res) {
          // if request error...
          if(err) {
            node.log('error: ' + err.message);
          }

          // else, if response ok or webhook not found...
          else if(res && res.statusCode && (res.statusCode == 204 || res.statusCode == 404)) {
            if(typeof id !== 'undefined') {
              node.webhookId = null;
              node.webhookObj = null;
            }
          }

          // else, response not ok...
          else {
            var statusCode = res && res.statusCode && res.statusCode > 1 ? res.statusCode : 500;
            var errorMsg = 'response code "' + res.statusCode + '" received removing webhook for "' + node.name + '"';
            node.log('error: ' + errorMsg);
          }

        });
      }
    };

    // create new webhook
    node.createWebhook = function() {

      // if Webex Teams api token and webhook target are defined...
      if(node.profileConfig.credentials && node.profileConfig.credentials.token && node.webhookTarget) {

        // clone module request template
        var _webhookReq = _.clone(webhookReq);

        // set credentials for request object
        _webhookReq.auth = { bearer: node.profileConfig.credentials.token };

        // set body for request object
        _webhookReq.body = {
          name: 'NODE-RED:' + node.id,
          targetUrl: node.webhookTarget,
          resource: node.resource,
          event: 'all'
        };

        // send request to add a webhook
        request.post(_webhookReq, function(err, res) {
          // if request error...
          if(err) {
            node.warn('error: ' + err.message);
          }

          // else, if response ok...
          else if(res && res.statusCode && res.statusCode == 200 && typeof res.body === 'object') {
            node.webhookObj = res.body;
            node.webhookId = res.body.id;
            node.log('Created webhook for: ' + node.webhookTarget);
          }

          // else, response not ok...
          else {
            var statusCode = res && res.statusCode && res.statusCode > 1 ? res.statusCode : 500;
            var errorMsg = 'response code "' + res.statusCode + '" received adding webhook for "' + node.name + '"';
            node.error('error: ' + errorMsg);
          }

        });
      }

      // else, Webex Teams api token not defined...
      else {
        var errorMsg = 'api token or webhookTarget not defined';
        node.error('error: ' + errorMsg);
      }
    };

    // indicate traffic via status
    node.reqPing = function() {
      // increment request counter
      node.reqCount++;

      var timeout;

      var online = {
        fill: 'blue',
        shape: 'ring',
        text: 'Webex Teams Webhook: online (' + node.reqCount + ')'
      };

      var recieving = {
        fill: 'blue',
        shape: 'dot',
        text: 'Webex Teams Webhook: online (' + node.reqCount + ')'
      };

      if(node.reqReceiving) {
        clearTimeout(timeout);
      } else {
        node.status(recieving);
      }

      node.reqReceiving = true;

      timeout = setTimeout(function() {
        node.reqReceiving = false;
        node.status(online);
      }, 200);
    };

    node.processError = function(err, req, res, next) {
      node.log(err);
      res.sendStatus(500);
    };

    node.processRequest = function(req, res, next) {
      var msg = {};
      msg.payload = {};
      msg.error = null;

      // if req and res
      if(req && res) {

        // if properties in body are valid...
        if(typeof req.body === 'object' && req.body.resource && req.body.event) {

          // if webhook matches resource and event
          if(req.body.resource === node.resource && req.body.event === node.event) {

            // attach headers to msg
            msg.headers = req.headers;

            // attach message body
            msg.payload = req.body;

            // ping the metric counter
            node.reqPing();

            // send msg
            node.send(msg);
          }
        }

        // send 200 OK
        res.status(200).send('OK');
      }

      // else next
      else if(typeof next === 'function') {
        node.log('error: could not process request');
        next();
      }

      // else
      else {
        node.log('error: could not process request');
      }
    };

    // cleanup node on deploy
    node.on('close',function() {
      var node = this;

      // remove route
      RED.httpNode._router.stack.forEach(function(route,i,routes) {
        if (route.route && route.route.path.match(/^\/node_red_contrib_spark_.*/)) {
          routes.splice(i,1);
          node.log('Removed existing route for: ' + node.pathname);
        }
      });

      // remove webhook
      node.removeWebhook();
    });

    // create route for this node
    RED.httpNode.post('/' + node.basename, jsonParser, node.processRequest, node.processError);
    node.log('Created new route for: ' + node.pathname);

    // remove webhooks with same Target Url and then create a new one...
    node.removeWebhooks(function(err) {
      // create webhook
      node.createWebhook();
    });

  }

  if(RED.settings.httpNodeRoot !== false) {
    RED.httpNode.get('/webhook/cisco_spark_v1.json', function(req, res) {
      var swaggerPath = path.resolve(__dirname, './webhook/cisco_spark_v1.json');
      fs.readFile(swaggerPath, function(err, data) {
        if(err) {
          res.set('Content-Type', 'text/javascript').send('{ "error": "' + err.message + '", "message": "Error reading webhook/cisco_spark_v1.json" }');
        } else {
          res.set('Content-Type', 'text/javascript').send(data);
        }
      });
    });

  } else {
    this.warn('httpNodeRoot is disabled in node-red settings');
  }

  RED.nodes.registerType('Webex Teams Webhook', SparkWebhookNode);
};


spark/spark-payload-parser.js
module.exports = function(RED) {
  'use strict';
  var _ = require('lodash');

  function SparkParserNode(n) {
    RED.nodes.createNode(this, n);

    var node = this;

    node.parsers = n.parsers;

    // return value
    function getPayloadKeyValue(payload, key) {
      // validate payload is object
      if(typeof payload === 'object') {

        var value = '';

        // search msg.payload.data
        if(payload.hasOwnProperty('data') && payload.data.hasOwnProperty(key)) {
          value = payload.data[key];
        }

        // search msg.payload
        else if(payload.hasOwnProperty(key)) {
          value = payload[key];
        }

        if(value != '') {
          return value;
        }
      }
    }

    function processMsg(msg, parsers) {
      var newMsg = _.cloneDeep(msg);
      var newPayload = {};

      if(parsers && parsers instanceof Array && parsers.length > 0) {
        _.forEach(parsers, function(parser) {
          if(parser.hasOwnProperty('key') && parser.key !== '') {
            if(!parser.hasOwnProperty('as') || (parser.hasOwnProperty('as') && parser.as === '')) {
              parser.as = parser.key;
            }
            newPayload[parser.as] = getPayloadKeyValue(newMsg.payload, parser.key);
          }
        });
      }

      if(newPayload !== {}) {
        newMsg.payload = newPayload ;

        return newMsg;
      }
    }

    // input event
    node.on('input', function(msg) {

      // if input is valid
      if(msg && typeof msg === 'object' && msg.hasOwnProperty('payload')) {

        // create new message and preserve properties that this node is not manipulating
        var newMsg = _.cloneDeep(msg);

        // if payload is object
        if(typeof newMsg.payload === 'object') {
          newMsg = processMsg(newMsg, node.parsers);

          // if parser returned value
          if(newMsg) {
            node.send(newMsg);
          }
        }
      }

      // else input is invalid
      else {
        node.warn('invalid input');
      }
    });
  }

  RED.nodes.registerType('Webex Teams Payload Parser', SparkParserNode);
};


spark/spark-api.js
module.exports = function(RED) {
  'use strict';
  var SwaggerClient = require('swagger-client');
  var path = require('path');
  var fs = require('fs');
  var _ = require('lodash');

  var swaggerClient = null;
  var swaggerUrl = null;

  // create swagger client
  function createClient(url, callback) {
    // if swagger client already exists...
    if(swaggerClient && swaggerUrl) {
      return callback(null);
    }

    // if swagger url has changed...
    if(swaggerClient && url !== swaggerUrl) {
      swaggerUrl = null;
      swaggerClient = null;
    }

    // update swagger url
    swaggerUrl = url;

    // create swagger client
    swaggerClient = new SwaggerClient({
      url: swaggerUrl,
      success: function() {
        callback(null);
      }, function(err) {
        callback(new Error('error'));
      }
    });
  }

  // send api request
  function sendRequest(msg, resource, method, params, opts, token, processResponse, processError) {
    if(token && swaggerClient) {
      swaggerClient.clientAuthorizations.add('apiKey', new SwaggerClient.ApiKeyAuthorization('Authorization', 'Bearer ' + token, 'header'));
      swaggerClient[resource][method](params, opts, function(response) {
        processResponse(msg, response);
      }, function(err) {
        processError(msg, err);
      });
    } else {
      node.error('webex teams api token or swagger client invalid or not defined');
    }
  }

  function SparkApiNode(n) {
    RED.nodes.createNode(this, n);

    var node = this;

    node.profileConfig = RED.nodes.getNode(n.profileConfig);
    node.apiUrl = n.apiUrl;
    node.resource = n.resource;
    node.method = n.method;
    node.id = n.id;

    node.reqCount = 0;
    node.reqReceiving = false;

    // set node status visual indicators
    function setNodeStatus(state) {
      var status = {};
      switch(state) {
        case 'warning':
          status = {
            fill: 'yellow',
            shape: 'dot',
            text: 'Webex Teams API: warning'
          };
          break;
        case 'error':
          status = {
            fill: 'red',
            shape: 'dot',
            text: 'Webex Teams API: error'
          };
          break;
        case 'input_error':
          status = {
            fill: 'red',
            shape: 'dot',
            text: 'Webex Teams API: invalid input'
          };
          break;
        default:
          status = {
            fill: 'blue',
            shape: 'ring',
            text: 'Webex Teams API: ready'
          };
          break;
      }

      node.status(status);
    }

    // indicate traffic via status
    function reqPing() {
      // increment request counter
      node.reqCount++;

      var timeout;

      var online = {
        fill: 'blue',
        shape: 'ring',
        text: 'Webex Teams API: online (' + node.reqCount + ')'
      };

      var recieving = {
        fill: 'blue',
        shape: 'dot',
        text: 'Webex Teams API: online (' + node.reqCount + ')'
      };

      if(node.reqReceiving) {
        clearTimeout(timeout);
      } else {
        node.status(recieving);
      }

      node.reqReceiving = true;

      timeout = setTimeout(function() {
        node.reqReceiving = false;
        node.status(online);
      }, 200);
    }

    function processResponse(msg, response) {
      reqPing();

      var newMsg = {};
      newMsg.topic = node.resource + '.' + node.method;

      // handle object response
      if(response && typeof response === 'object') {

        // capture headers
        if(response.hasOwnProperty('headers') && typeof response.headers === 'object') {
          newMsg.headers = response.headers;
        }

        // capture status
        if(response.hasOwnProperty('status')) {
          newMsg.status = response.status;
        } else {
          newMsg.status = 500;
        }

        // if response is from a delete...
        if(newMsg.status === 204) {
          newMsg.payload = {};

          // send message
          node.send(newMsg);

          return null;
        }

        // if response.data
        if(response.hasOwnProperty('data')) {
          var data = response.data;

          // if empty response...
          if(typeof data === 'string' && (data === '' || data === '{}')) {
            // show warning
            node.warn('error processsing response');

            // set node status
            setNodeStatus('warning');

            return null;
          }

          // if response.data is valid...
          if(typeof data === 'string') {

            // attempt parsing of json
            try {
              data = JSON.parse(data);
            }
            catch(err) {
              // show warning
              node.warn('error: ' + err.message || 'undefined');

              // set node status
              setNodeStatus('warning');

              return null;
            }

            // if response is array of items
            if(typeof data === 'object' && data.hasOwnProperty('items') && data.items instanceof Array) {

              // if as multiple messages
              if(true) {
                // define _msgid
                newMsg._msgid = RED.util.generateId();

                // define messages
                var newMsgArray = _.map(data.items, function(item, index) {
                  var msg = _.cloneDeep(newMsg);

                  msg.payload = item;
                  msg.parts = {
                    "id": msg._msgid,
                    "type": "array",
                    "index": index,
                    "count": data.items.length
                  };
                  return msg;
                });

                // send message
                node.send([newMsgArray]);
              }

              // else, as single message
              else {
                // define _msgid
                newMsg._msgid = msg._msgid;

                // define paypload
                newMsg.payload = data.items;

                // send message
                node.send(newMsg);
              }

            }


            // else, response is single item
            else {
              newMsg.payload = data;

              // send message
              node.send(newMsg);
            }
          }
        }

        // else, response.data does not exist...
        else {
          // show warning
          node.warn('response has no content');

          // set node status
          setNodeStatus('warning');
        }
      }

      else {
        node.warn('response invalid or not received');

        // set node status
        setNodeStatus('warning');
      }
    }

    function processError(msg, response) {
      if(response && typeof response === 'object') {
        var status;
        var data;
        var errMessage;
        var trackingId;

        // capture status
        if(response.hasOwnProperty('status')) {
          status = response.status;
        } else {
          status = 500;
        }

        // capture response body
        if(response.hasOwnProperty('data')) {
          data = response.data;

          // if response.data is valid...
          if(typeof data === 'string') {

            // attempt parsing of json
            try {
              data = JSON.parse(data);
            }
            catch(err) {
              node.error('error ' + status + ': ' + err.message || 'undefined');

              // set node status
              setNodeStatus('error');

              return null;
            }

            if(data.hasOwnProperty('trackingId')) {
              trackingId = data.trackingId;
            } else {
              trackingId = 'unknown';
            }

            if(data.hasOwnProperty('message')) {
              errMessage = data.message;
            } else {
              errMessage = 'unknown';
            }

            node.error('error ' + status + ': ' + errMessage + ' | trackingId: ' + trackingId);
          }
        } else {
          node.error('error ' + status + ': undefined error');
        }
      } else {
        node.error('error: unknown API response');
      }

      // set node status
      setNodeStatus('error');
    }

    // start node if profileConfig is defined
    if(node.profileConfig) {

      // set default status
      setNodeStatus();

      // create client
      createClient(node.apiUrl, function(err) {

        // if error creating client
        if(err) {
          // show warning
          node.warn('invalid input');

          // set node status
          setNodeStatus('input_error');

          return;
        }

        // else, client created
        else {

          // create input event
          node.on('input', function(msg) {

            var opts = {
              requestContentType: 'application/json',
              responseContentType: 'application/json'
            };

            var params = {};

            // if input is non empty string
            if(msg && typeof msg.payload === 'string' && msg.payload.length > 0) {
              // atempt parsing of json
              try {
                params = JSON.parse(msg.payload);
                if(params instanceof Array) {
                  // show warning
                  node.warn('invalid input');

                  // set node status
                  setNodeStatus('input_error');

                  return;
                }
              }
              catch(err) {
                // show warning
                node.warn(err.message);

                // set node status
                setNodeStatus('input_error');

                return;
              }
            }

            // else if input is object
            else if(msg && typeof msg.payload === 'object' && !(msg.payload instanceof Array)) {
              params = msg.payload;
            }

            // else input is invalid
            else {
              // show warning
              node.warn('invalid input');

              // set node status
              setNodeStatus('input_error');

              return;
            }

            // Send Spark API Request
            sendRequest(msg, node.resource, node.method, params, opts, node.profileConfig.credentials.token, processResponse, processError);
          });
        }
      });
    }
  }

  if(RED.settings.httpNodeRoot !== false) {
    RED.httpNode.get('/swagger-client-web.js', function(req, res) {
      var clientPath = path.resolve(__dirname, './swagger-client-web.js');
      fs.readFile(clientPath, function(err, data) {
        if(err) {
          res.set('Content-Type', 'text/javascript').send('{ "error": "' + err.message + '", "message": "Error reading swagger-client-web.js" }');
        } else {
          res.set('Content-Type', 'text/javascript').send(data);
        }
      });
    });

    RED.httpNode.get('/api/cisco_spark_v1.json', function(req, res) {
      var apiPath = path.resolve(__dirname, './api/cisco_spark_v1.json');
      fs.readFile(apiPath, function(err, data) {
        if(err) {
          res.set('Content-Type', 'text/javascript').send('{ "error": "' + err.message + '", "message": "Error reading api/cisco_spark_v1.json" }');
        } else {
          res.set('Content-Type', 'text/javascript').send(data);
        }
      });
    });
  } else {
    this.warn('httpNodeRoot is disabled in node-red settings');
  }

  RED.nodes.registerType('Webex Teams API', SparkApiNode);
};
